src/irmin-pack/checks.ml

```diff
@@ -0,0 +1,225 @@
+
+  let toplevel root = [ flip root; lower root; upper1 root; upper0 root ]
+end
+
+module Make
+    (Conf : Config.S)
+    (M : Irmin.Metadata.S)
+    (C : Irmin.Contents.S)
+    (P : Irmin.Path.S)
+    (B : Irmin.Branch.S)
+    (H : Irmin.Hash.S)
+    (Node : Irmin.Private.Node.S
+              with type metadata = M.t
+               and type hash = H.t
+               and type step = P.step)
+    (Commit : Irmin.Private.Commit.S with type hash = H.t) =
+struct
+  module Index = Pack_index.Make (H)
+
+  type size = Bytes of int [@@deriving irmin]
+
+  let detect_layered_store ~root =
+    root |> Layout.toplevel |> List.exists (fun path -> IO.exists path)
+
+  let detect_pack_layer ~layer_root = Layout.dict layer_root |> IO.exists
+
+  let read_flip ~root =
+    let path = Layout.flip root in
+    match IO.exists path with
+    | false -> Lwt.return_none
+    | true ->
+        IO_layers.IO.v path >>= fun t ->
+        (IO_layers.IO.read_flip t >|= function
+         | true -> `Upper1
+         | false -> `Upper0)
+        >>= fun a ->
+        IO_layers.IO.close t >|= fun () -> Some a
+
+  (** Read basic metrics from an existing store. *)
+  module Quick_stats = struct
+    type io = { size : size; offset : int64; generation : int64 }
+    [@@deriving irmin]
+
+    type files = { pack : io option; branch : io option; dict : io option }
+    [@@deriving irmin]
+
+    type files_layer = {
+      flip : [ `Upper1 | `Upper0 ] option;
+      lower : files;
+      upper1 : files;
+      upper0 : files;
+    }
+    [@@deriving irmin]
+
+    type files_store = Layered of files_layer | Simple of files
+    [@@deriving irmin]
+
+    type t = { hash_size : size; files : files_store } [@@deriving irmin]
+
+    let with_io : type a. string -> (IO.t -> a) -> a option =
+     fun path f ->
+      match IO.exists path with
+      | false -> None
+      | true ->
+          let io =
+            IO.v ~fresh:false ~readonly:true ~version:(Some current_version)
+              path
+          in
+          let a = f io in
+          IO.close io;
+          Some a
+
+    let io path =
+      with_io path @@ fun io ->
+      let offset = IO.offset io in
+      let generation = IO.generation io in
+      let size = Bytes (IO.size io) in
+      { size; offset; generation }
+
+    let v_simple ~root =
+      let pack = Layout.pack root |> io in
+      let branch = Layout.branch root |> io in
+      let dict = Layout.dict root |> io in
+      { pack; branch; dict }
+
+    let v_layered ~root =
+      match detect_layered_store ~root with
+      | false -> Lwt.return (Simple (v_simple ~root))
+      | true -> (
+          Logs.app (fun f -> f "Layered store detected");
+          read_flip ~root >|= fun flip ->
+          Layout.toplevel root
+          |> List.tl
+          |> List.map (fun root -> v_simple ~root)
+          |> function
+          | [ lower; upper1; upper0 ] -> Layered { flip; lower; upper1; upper0 }
+          | _ -> assert false)
```

Comment:
Looks like it might be better expressed with some duplication:

```suggestion
          let lower = v_simple ~root:Layout.lower
          and upper1 = v_simple ~root:Layout.upper1
          and upper2 = v_simple ~root:Layout.upper2
          in
          Layered { flip; lower; upper1; upper0 }
```

(The precise layout of that list is an invariant that may be moved elsewhere in future.)

Of course, you could define a `Triple` module if you wanted :wink: